crates/turborepo-lib/src/shim.rs

```diff
@@ -227,114 +226,156 @@ pub struct RepoState {
+            // More than one potential root. See if we can stop at the first.
+            // This is a performance optimization. We could remove this case,
+            // and set the mode properly in the else and it would still work.
+            } else if check_roots[0].package_has_workspaces {
+                // If the closest one has workspaces then we stop there.
+                let local_turbo_state = LocalTurboState::infer(check_roots[0].path.to_path_buf());
+                return Ok(Self {
+                    root: check_roots[0].path.to_path_buf(),
+                    mode: RepoMode::MultiPackage,
+                    local_turbo_state,
+                });
 
-            if is_workspace {
-                let local_turbo_state = LocalTurboState::infer(dir);
+            // More than one potential root.
+            // Closest is not RepoMode::MultiPackage
+            // We attempt to prove that the closest is a workspace of a parent.
+            // Failing that we just choose the closest.
+            } else {
+                for ancestor_infer in &check_roots[1..] {
+                    if ancestor_infer.is_workspace_root_of(check_roots[0].path.to_path_buf()) {
+                        let local_turbo_state =
+                            LocalTurboState::infer(check_roots[0].path.to_path_buf());
+                        return Ok(Self {
+                            root: ancestor_infer.path.to_path_buf(),
+                            mode: RepoMode::MultiPackage,
+                            local_turbo_state,
+                        });
+                    }
+                }
 
+                // We have eliminated RepoMode::MultiPackage as an option.
+                // We must exhaustively check before this becomes the answer.
+                let local_turbo_state = LocalTurboState::infer(check_roots[0].path.to_path_buf());
                 return Ok(Self {
-                    root: dir.to_path_buf(),
-                    mode: RepoMode::MultiPackage,
+                    root: check_roots[0].path.to_path_buf(),
+                    mode: RepoMode::SinglePackage,
                     local_turbo_state,
                 });
             }
         }
 
-        // Finally, if we don't detect any workspaces, go to the first `package.json`
-        // and use that in single package mode.
-        let root = first_package_json_dir
-            .ok_or_else(|| {
-                anyhow!(
-                    "Unable to find `{}` or `package.json` in current path",
-                    TURBO_JSON
-                )
-            })?
-            .to_path_buf();
-
-        let local_turbo_state = LocalTurboState::infer(&root);
-        Ok(Self {
-            root,
-            mode: RepoMode::SinglePackage,
-            local_turbo_state,
-        })
+        // If we're here we didn't find a valid root.
+        return Err(anyhow!("Root could not be inferred."));
```

Comment:
Can drop the `return`

Reply:
Clippy didn't catch this one for some reason; I'm guessing because of the macro?